### DISCUSSION TITLE
Update one table row with the maximum field value of several rows of another table

### DISCUSSION THREAD

**Discussion by gerhardmpl** [Budibase/budibase#4465] [Help] [answered]
Only 24h into Budibase and already really impressed.

Currently doing a PoC to create an app for our Information Security Management System (ISMS). I need some help with how to update a table with the maximum value of field from several rows when one row is changed.

There is a table that holds the company’s processes (e.g. order to cash) and a table that holds the information used in the processes (e.g customer order, customer invoice, delivery note). Table processes is linked to table information by a "many to many" relationship.

As part of the business impact analysis (BIA), every information is evaluated regarding confidentiality, integrity, and availability. I have three other tables that hold the definition of what that means for the company (e.g. level 1 with name = "low" or level = 4 with name = "very high"). Table information is linked to tables confidentiality, integrity, and availability by a "on to many" relationship.

Getting to the part where I need help, every information is than related to the service (application, cloud service) that stores or processes that information. Table information is linked to table service by a "many to many" relationship since information can be processed in different services. How can I pass on the maximum confidentiality, integrity or availability of related information rows to the respective fields in the services row (e.g. if the availability for customer order has level 2 with name "medium" and customer invoice has level 3 with name "high", how to I update the maximum required availability for the "ERP service" to level 3 with name "high"?

I assume it is done with automation, most likely with a JavaScript SQL statement, but maybe there is another option I am not aware of. Thanks for any suggestion.

If it helps I could attache a sceleton data structure with examples.

**Comment by mjashanks** [comment]
Hey @gerhardmpl - thanks & welcome!

Your data structure would really help.

Question - once you get the "Final Result" (i.e. "high"),  what do you want to do with that result? Just display on screen?

**Comment by gerhardmpl** [comment]
Hello @mjashanks and thanks for the swift reaction. I will attach the data structur later this day.

I need the result as an evaluatable value because in the risk analysis that follows the BIA, the services must pass on the maximum confidentiality, integrity or availability to linked assets (same logic but different tables). I was wondering if a formular type field could help here, but you already answered that in #4474

**Comment by gerhardmpl** [comment]
@mjashanks attached please find the data structure for the BIA part. 

Table _information_ plays a central role for the business impact analysis. Thats where you select the protection requirements regarding confidentiality, integrity or availability and link the information to the information processing services. 

From the example you see that the overall protection requirements for the ERP service is medium, high, high.
- If you add a new row _delivery note_ with high, medium, low and link it to the ERP service, the overall requirement would change to high, high, high
- If you instead update _invioce (from supplier)_ to medium, high, low, the overall requirement would change to medium, high, medium.

As a result, the row _ERP Service_ in table _services_ needs to be updated, when an information row is changed regarding the protection requirements.

The main idea is that the maximum protection requirement of an individual information for one of the security goals (confidentiality, integrity or availability) determines the corresponding protection requirement of the IT service. Table _confidentiality_, _integrity_ and _availability_ are used to flexibly classify criteria (avoiding static column type Options).

I know this is a very specific use case, so I appreciate any help or pointers in the right direction

[ISMS Portal-export-1644852296732.txt](https://github.com/Budibase/budibase/files/8061943/ISMS.Portal-export-1644852296732.txt)

**Reply by mjashanks** [reply]
Thanks for this - i think I understand.

The first thing that sticks out to me is that you seem to be using relationships alot more than you need to. For example...

- Your `Information` table has a column called `Confidentiality` - a relationship to the confidentiality table
- Your `Confidentiality` table has 4 values - `low`, `medium`, `high` and `very high`

Can the `Confidentiality` **column** simply be a text field that stores "low" / "medium" etc ? Doing it this way would mean that you could just use a simple formula to calculate the final result.

**Comment by gerhardmpl** [comment]
Using relationship is more complex, I agree. But it would allow me to easily change the numbers of levels,  have a number field for numeric (non-alphabetic) sorting (avoiding an option like ```1 - low``` where ```low``` is less distracting) and add an explanation field for what ```low``` means to the organisation (left that content out in the examples but the field is there (type should be ```Long Form Text``` and not ```Text```)).

How would I query the ```information``` table from a formula field in the ```services``` table to find out which related multiple information rows have what ```Confidentiality``` level to only pass on the maximum level to the service row?

Coming from a more traditional developer role (which might show through my approach), it is great fun to dip my toe into the Low Code concepts.

**Reply by mjashanks** [reply]
You could use a `Lookups` table to store `high` , `medium` and `low` - and include an `order` column. You can use this table to load options into a dropdown. I describe this here: https://github.com/Budibase/budibase/discussions/4418#discussioncomment-2156404

Note that your columns would still be `Text`, but they just display a dropdown of options.

**Comment by gerhardmpl** [comment]
@mjashanks trying to simplify for the time beeing, I dropped the relationship for CIA (confidentiality, integrity, availability) and use Options and Formular fields. To update the ```services``` table when a row in the ```information``` table is created, updated or deleted, I most likely have to go the Automation route.

1. Automation gets triggerd on ```Update Row``` of ```Information```
2. Add a ```JS Scripting``` Action to query the ```services``` relationship from the ```trigger.row``` and find the related services to update
3. Add a ```Update Row``` Action to update the related rows in the ```services``` table

Could that work and how can I query the relationship (e.g. ```service```) in the ```trigger.row``` binding of step1, so that I can read the related services (e.g. _ERP Service_ and _CRM Service_) and return a list of rows to step 2. Unfortunatly, I cannot use an ```query row``` Action for that I understand.

**Reply by mjashanks** [reply]
The Javascript automation step is for simple data manipulation. You cannot do any querying from here.

After that, you need the ability to query the `services` table by `_id` - we are adding this at the end of this month.

**Comment by gerhardmpl** [comment]
While I can query the ```services``` relationship from the ```trigger.row``` binding in the ```JS Scripting``` Action, I can not figure out how to continue with the result (this is an example for one information row that is linked to two services rows through relationsship):

```
{
  "success": true,
  "value": [
    "ro_ta_1fb6cb1528e34b2e9293670ac3a4ab38_165ebdfaaee3471b93f7e052fb383435",
    "ro_ta_1fb6cb1528e34b2e9293670ac3a4ab38_bc5b13a6fedd47cd8a4a0fa4522aaefc"
  ]
}
```

My hope was to access the ```services``` array through something like ```trigger.row.services.0.Auto ID```, but this looks more like an object reference for an internal bbdb table.

Sorry for mumbling to myself. I am afraid this discussion can be closed.

**Reply by mjashanks** [reply]
Those ids refer to a budibase internal id - named `_id`. Unfortunately the `Query rows` does not have the option to query by `_id`, but it's something that we will have available by the end of this month.

**Comment by gerhardmpl** [comment]
Refering to your answer in  #4514 ```Query Rows``` is the only way: does ```Query Rows``` filter allow for multiple selections (like the two internal ```_id``` above)? 

I have found a workaround to get the ```Auto ID``` values in the ```JS Scripting``` Action, but do not know how to hand that over to the next ```Query Rows``` Automation step and loop through the response for filtering (assuming that e.g. ```steps.1.response``` is an array).

Thank you for bearing with me and my Beginners questions.

**Comment by gerhardmpl** [comment, marked as answer]
I close this discussion for the time beeing and go back to the drawing board. While I realy like the relationship functionality, the available options to manipulate or query data cross (internal bbdb) tables especially when using automation (escaping the binding context) adds way to much complexity for the moment. Thanks again for your support.
